src/sys/unix/uds.rs

```diff
@@ -0,0 +1,298 @@
+use crate::sys::unix::net::new_socket;
+
+use std::ascii;
+use std::cmp::Ordering;
+use std::ffi::OsStr;
+use std::fmt;
+use std::io;
+use std::mem;
+use std::os::unix::net::{UnixDatagram, UnixListener, UnixStream};
+use std::os::unix::prelude::*;
+use std::path::Path;
+
+/// An address associated with a mio specific unix socket.
+///
+/// This is implemented instead of imported from [`net::SocketAddr`] because
+/// there is no way to create a [`net::SocketAddr`]. One must be returned by
+/// [`accept`], so this is returned instead.
+///
+/// [`net::SocketAddr`]: std::os::unix::net::SocketAddr
+/// [`accept`]: #method.accept
+pub struct SocketAddr {
+    addr: libc::sockaddr_un,
+    len: libc::socklen_t,
+}
+
+enum AddressKind<'a> {
+    Unnamed,
+    Pathname(&'a Path),
+    Abstract(&'a [u8]),
+}
+
+pub fn connect_stream(path: &Path) -> io::Result<UnixStream> {
+    let socket = new_socket(libc::AF_UNIX, libc::SOCK_STREAM)?;
+    let (raw_addr, raw_addr_length) = socket_addr(path)?;
+    let raw_addr = &raw_addr as *const _ as *const _;
```

Comment:
Why is the second cast to `*const _` necessary?

Reply:
The casts are `&` -> `*const libc::sockaddr_un` -> `*const libc::sockaddr`. The `*const` casts can be elided, but I could put them back in so that it is more explicit.

I don't think a direct cast can happen from `&` to `*const libc::sockaddr`.

Reply:
hmm, take it or leave it, but i might just put the actual types in:

```rust
    let raw_addr = &raw_addr as *const libc::sockadder_un as *const libc::sockaddr;
```
imo this is clearer.

Reply:
Yea I like it too. It's added to the rest of the elided casts as well.